File: magellan-library/src/main/java/com/wealthfront/magellan/transitions/CircularRevealTransition.java
package com.wealthfront.magellan.transitions;

import android.animation.Animator;
import android.animation.AnimatorListenerAdapter;
import android.graphics.Rect;
import android.view.View;
import android.view.ViewAnimationUtils;
import android.view.ViewGroup;

import com.wealthfront.magellan.Direction;
import com.wealthfront.magellan.NavigationType;

public class CircularRevealTransition implements Transition {

  private final View clickedView;

  public CircularRevealTransition(View clickedView) {
    this.clickedView = clickedView;
  }

  @Override
  public void animate(
      View from, View to, NavigationType navType, Direction direction, final Callback callback) {
    int[] clickedViewCenter = getCenterClickedView((ViewGroup) from);
    int circularRevealCenterX = clickedViewCenter[0];
    int circularRevealCenterY = clickedViewCenter[1];
    float finalRadius = (float) Math.hypot(to.getWidth(), to.getHeight());

    if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.LOLLIPOP) {
      Animator anim =
          ViewAnimationUtils.createCircularReveal(to, circularRevealCenterX,
              circularRevealCenterY, 0, finalRadius);
      anim.addListener(new AnimatorListenerAdapter() {
        @Override
        public void onAnimationEnd(Animator animation) {
          callback.onAnimationEnd();
        }
      });
      anim.start();
    } else {
      callback.onAnimationEnd();
    }
  }

  private int[] getCenterClickedView(ViewGroup from) {
    Rect clickedViewRect = new Rect();
    clickedView.getDrawingRect(clickedViewRect);
    from.offsetDescendantRectToMyCoords(clickedView, clickedViewRect);
    return new int[] {(int) clickedViewRect.exactCenterX(), (int) clickedViewRect.exactCenterY()};
  }

}

File: magellan-library/src/main/java/com/wealthfront/magellan/transitions/CircularRevealTransition.kt
package com.wealthfront.magellan.transitions

import android.animation.Animator
import android.animation.AnimatorListenerAdapter
import android.graphics.Rect
import android.os.Build
import android.view.View
import android.view.ViewAnimationUtils
import android.view.ViewGroup
import androidx.annotation.IdRes
import androidx.interpolator.view.animation.FastOutSlowInInterpolator
import com.wealthfront.magellan.Direction
import kotlin.math.hypot

@Suppress("FunctionName")
public fun CircularRevealTransition(clickedView: View): MagellanTransition {
  return CircularRevealTransition(clickedView.id)
}

/**
 * A [MagellanTransition] that reveals the next screen circularly outward from the middle of the
 * [clickedViewId].
 *
 * @property clickedViewId the id of the view on which this circular reveal is centered
 */
private class CircularRevealTransition(@IdRes private val clickedViewId: Int) : MagellanTransition {

  private var animator: Animator? = null

  override fun interrupt() {
    animator?.end()
  }

  override fun animate(
    from: View?,
    to: View,
    direction: Direction,
    onAnimationEndCallback: () -> Unit
  ) {
    if (Build.VERSION.SDK_INT < Build.VERSION_CODES.LOLLIPOP) {
      onAnimationEndCallback()
      return
    }

    val clickedViewContainer = if (direction == Direction.FORWARD) from else to
    val viewToReveal = if (direction == Direction.FORWARD) to else from!!
    val clickedViewCenter = getCenterClickedView(clickedViewContainer as ViewGroup)
    val circularRevealCenterX = clickedViewCenter[0]
    val circularRevealCenterY = clickedViewCenter[1]
    val revealedRadius = hypot(viewToReveal.width.toDouble(), viewToReveal.height.toDouble()).toFloat()
    val finalRadius = if (direction == Direction.FORWARD) revealedRadius else 0f
    val startRadius = if (direction == Direction.FORWARD) 0f else revealedRadius
    animator = ViewAnimationUtils.createCircularReveal(
      viewToReveal, circularRevealCenterX,
      circularRevealCenterY, startRadius, finalRadius
    ).apply {
      addListener(object : AnimatorListenerAdapter() {
        override fun onAnimationEnd(animation: Animator) {
          animator = null
          onAnimationEndCallback()
        }
      })
      interpolator = FastOutSlowInInterpolator()
    }.also { it.start() }
  }

  private fun getCenterClickedView(clickedViewContainer: ViewGroup): IntArray {
    val clickedView = clickedViewContainer.findViewById<View>(clickedViewId)
    val clickedViewRect = Rect()
    clickedView.getDrawingRect(clickedViewRect)
    clickedViewContainer.offsetDescendantRectToMyCoords(clickedView, clickedViewRect)
    return intArrayOf(
      clickedViewRect.exactCenterX().toInt(),
      clickedViewRect.exactCenterY().toInt()
    )
  }
}
